### PR TITLE
Bump gradle-version, dependencies, support JDK 11

### PR DESCRIPTION
**Pull request by ank27**
### Summary of changes
Dependencies updated
- Android gradle plugin to 7.0.4.
- Gradle version to 7.0.2.
- Gradle licence plugin to 0.8.80.
- Jacoco to 0.8.7.
- Kotlin gradle plugin to 1.5.31.

**Since we adapt to Gradle plugin 7.0, JDK 11 is now required to run the project.** 

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>Update android gradle plugin to v7.0.4, gradle version to v7.0.2, Gradle licence plugin to 0.8.80, Kotlin gradle plugin to 1.5.31, Jacoco to 0.8.7</changelog>`.
 - [x] If this PR is a `v10.4` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.4` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: LICENSE.md [hunk 1]
```diff
@@ -12,6 +12,22 @@ Mapbox Maps for Android version 10.0 ('Mapbox Maps Android SDK') or higher must
...
 ## Additional Mapbox Maps Android licenses
+Mapbox Maps Android uses portions of the Activity.
+
+URL: [https://developer.android.com/jetpack/androidx/releases/activity#1.2.3](https://developer.android.com/jetpack/androidx/releases/activity#1.2.3)
+
+License: [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt)
+
+===========================================================================
+
+Mapbox Maps Android uses portions of the Activity.
```
Comment (yunikkk) [inline LICENSE.md:23]: Hmm why it says `Activity` ? Should be `saved state` I guess

**Review by yunikkk — APPROVED** [review]

**Comment by kiryldz** [comment]
@ank27 why should this be ported to stable v10.3? I'm judging based on a checkbox

**Comment by kiryldz** [comment]
Just pointing out that this PR will be split in 2: 
 - adding https://github.com/mapbox/android-sdk-versions-plugin - this should be ported to v10.3 if nothing critical comes out
 - bumping AGP, Gradle will be available starting from v10.4-beta.1

did I get it correct @ank27 ?

**Comment by ank27** [comment]
> Just pointing out that this PR will be split in 2:
> 
> * adding https://github.com/mapbox/android-sdk-versions-plugin - this should be ported to v10.3 if nothing critical comes out
> * bumping AGP, Gradle will be available starting from v10.4-beta.1
> 
> did I get it correct @ank27 ?

Yes, this pr is now related to gradle bump (for 10.4-beta.1), and versions plugin added to https://github.com/mapbox/mapbox-maps-android/pull/1123.

File: LICENSE.md (hunk 1 above)
Comment (ank27) [reply LICENSE.md:23]: looks like the `project` property for both the dependencies gives `Activity` value. this value comes from the license plugin. It should be `saved state` imo as well.
licence_generate

Comment (yunikkk) [reply LICENSE.md:23]: https://mvnrepository.com/artifact/androidx.savedstate/savedstate indeed, seems the docs are broken

File: LICENSE.md [hunk 2]
```diff
@@ -12,6 +12,22 @@ Mapbox Maps for Android version 10.0 ('Mapbox Maps Android SDK') or higher must
...
+
+URL: [https://developer.android.com/jetpack/androidx/releases/savedstate#1.1.0](https://developer.android.com/jetpack/androidx/releases/savedstate#1.1.0)
```
Comment (pengdev) [inline LICENSE.md:25]: do we actually use savedstate dependency in some of our dependencies, is it just introduced because of the version bump?

File: LICENSE.md [hunk 3]
```diff
@@ -20,6 +36,14 @@ License: [The Apache Software License, Version 2.0](http://www.apache.org/licens
...
 ===========================================================================
 
+Mapbox Maps Android uses portions of the Android AppCompat Library.
+
+URL: [https://developer.android.com/jetpack/androidx/releases/appcompat#1.3.0](https://developer.android.com/jetpack/androidx/releases/appcompat#1.3.0)
+
+License: [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt)
+
+===========================================================================
+
```
Comment (pengdev) [inline LICENSE.md:46]: I think our SDK doesn't use AppCompat? It's just our sample app?

File: android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/car/CarJavaInterfaceChecker.java [hunk 4]
```diff
@@ -5,10 +5,12 @@
 import androidx.car.app.Session;
 
 import com.mapbox.maps.MapInitOptions;
+import com.mapbox.maps.MapboxExperimental;
 import com.mapbox.maps.extension.androidauto.MapSurfaceReadyCallback;
 import com.mapbox.maps.extension.androidauto.OnMapScaleListener;
 import com.mapbox.maps.extension.androidauto.OnMapScrollListener;
 
+@MapboxExperimental
```
Comment (pengdev) [inline android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/car/CarJavaInterfaceChecker.java:13]: do we need to add this annotation?

File: gradle/lint.gradle [hunk 5]
```diff
@@ -3,6 +3,6 @@ android {
         checkReleaseBuilds false
         warningsAsErrors true
         xmlReport false
-        disable 'AllowBackup'
+        disable 'AllowBackup', 'UnusedResources'
```
Comment (pengdev) [inline gradle/lint.gradle:6]: I guess we want to warn unusedResources?

File: buildSrc/src/main/kotlin/Project.kt [hunk 6]
```diff
@@ -73,14 +73,14 @@ object Dependencies {
 }
 
 object Versions {
-  const val pluginAndroidGradle = "4.0.1"
-  const val pluginKotlin = "1.4.0"
-  const val pluginLicense = "0.8.5"
+  const val pluginAndroidGradle = "7.0.4"
+  const val pluginKotlin = "1.5.31"
```
Comment (pengdev) [inline buildSrc/src/main/kotlin/Project.kt:77]: Please also mention the Kotlin version change in the description and changelog.

File: gradle.properties [hunk 7]
```diff
@@ -24,6 +24,7 @@ kotlin.code.style=official
 # Work around Jetifier can't handle multi-release jars
 # https://github.com/bcgit/bc-java/issues/762
 # https://issuetracker.google.com/issues/172784894
-android.jetifier.blacklist=bcprov-jdk15on
+# android.jetifier.blacklist=bcprov-jdk15on
```
[the remark this answers is not in the document]
Comment (ank27) [reply gradle.properties:27]: Encountered below error when compiling.
`The option 'android.jetifier.blacklist' is deprecated. It was removed in version 7.0 of the Android Gradle plugin.
This property has been replaced by android.jetifier.ignorelist`

**Review by Chaoba — APPROVED** [review]

File: gradle.properties (hunk 7 above)
[the remark this answers is not in the document]
Comment (pengdev) [reply gradle.properties:27]: do we need to use something else? or is it safe to remove this line totally?

File: extension-style/api/extension-style.api [hunk 8]
```diff
@@ -3411,16 +3411,16 @@ public final class com/mapbox/maps/extension/style/sources/TileSet : java/util/H
 	public fun containsValue (Lcom/mapbox/bindgen/Value;)Z
 	public final fun containsValue (Ljava/lang/Object;)Z
 	public final fun entrySet ()Ljava/util/Set;
-	public final fun get (Ljava/lang/Object;)Ljava/lang/Object;
+	public final fun get (Ljava/lang/Object;)Lcom/mapbox/bindgen/Value;
```
Comment (kiryldz) [inline extension-style/api/extension-style.api:3414]: interesting, is this related to version bump?

**Comment by kiryldz** [comment]
@ank27 is it expected that we did pull so many new libraries (based on license file)? I remember we once did pull similar amount of libraries because common introduced dependency on play-services which increased our binary size a lot.

File: gradle.properties (hunk 7 above)
[the remark this answers is not in the document]
Comment (ank27) [reply gradle.properties:27]: Yes, we now use `android.jetifier.ignorelist`, which is at the line below this, actually i can remove the commented line to just keep the line that has `ignorelist` entry.

File: LICENSE.md (hunk 2 above)
Comment (ank27) [reply LICENSE.md:25]: I think these are introduced because we bumped the license version. Our sdk doesn't use these, it could be that some dependencies uses this dependency.

File: extension-style/api/extension-style.api (hunk 8 above)
Comment (ank27) [reply extension-style/api/extension-style.api:3414]: yeah looks like it. this is edited by kotlin binary compatibility check and verify-api also caught it as a binary breaking change (check the internal ticket regarding breaking changes). @pengdev any comment on this.

File: gradle/lint.gradle (hunk 5 above)
Comment (kiryldz) [reply gradle/lint.gradle:6]: agree with @pengdev here unless there is some very strong reason 😄

File: gradle/wrapper/gradle-wrapper.properties [hunk 9]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-all.zip
```
Comment (kiryldz) [inline gradle/wrapper/gradle-wrapper.properties:6]: @alexander-kulikovskii must be happy now 😄

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/animator/CameraAnimator.kt [hunk 10]
```diff
@@ -10,6 +11,7 @@ import com.mapbox.maps.plugin.animation.CameraAnimatorType
 /**
  * Base generic class for all camera animators.
  */
+@SuppressLint("Recycle")
```
Comment (kiryldz) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/animator/CameraAnimator.kt:14]: As a nit here and in other place below - can we mark it not for the whole class but for specific method?

File: sdk/api/sdk.api [hunk 11]
```diff
@@ -488,11 +488,6 @@ public abstract interface class com/mapbox/maps/renderer/OnFpsChangedListener {
 	public abstract fun onFpsChanged (D)V
 }
 
-public final class com/mapbox/maps/renderer/RenderHandlerThread$sam$i$java_lang_Runnable$0 : java/lang/Runnable {
-	public fun <init> (Lkotlin/jvm/functions/Function0;)V
-	public final synthetic fun run ()V
-}
-
```
Comment (kiryldz) [inline sdk/api/sdk.api:495]: pretty curious how did this happen

File: sdk/src/main/java/com/mapbox/maps/renderer/widget/WidgetPosition.kt [hunk 12]
```diff
@@ -4,21 +4,48 @@ import com.mapbox.maps.MapboxExperimental
 
 /**
  * Specifies widget position relative to the screen.
+ * @param horizontal position relative to the screen.
```
[the remark this answers is not in the document]
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/widget/WidgetPosition.kt:7]: interesting how did it pass before...

File: .gitignore [hunk 13]
```diff
@@ -28,3 +28,4 @@ release-docs.zip
 /tmp/checksum.txt
 api_compat_report/
 logs/
+**/src/main/assets/sdk_versions/
```
Comment (kiryldz) [inline .gitignore:31]: Let's mention this in PR description

File: android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/testing/CarJavaInterfaceChecker.java [hunk 14]
```diff
@@ -9,11 +9,13 @@
 import com.mapbox.maps.EdgeInsets;
 import com.mapbox.maps.MapInitOptions;
 import com.mapbox.maps.MapSurface;
+import com.mapbox.maps.MapboxExperimental;
 import com.mapbox.maps.ScreenCoordinate;
 import com.mapbox.maps.extension.androidauto.MapboxCarMap;
 import com.mapbox.maps.extension.androidauto.MapboxCarMapObserver;
 import com.mapbox.maps.extension.androidauto.MapboxCarMapSurface;
 
+@MapboxExperimental
```
Comment (kiryldz) [inline android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/testing/CarJavaInterfaceChecker.java:18]: why do this in this PR?

File: .gitignore (hunk 13 above)
Comment (kiryldz) [reply .gitignore:31]: @ank27 also I guess we need to explicitly mention in PR description and also in our `DEVELOPING.md` file that JDK is _required_ now.

File: LICENSE.md (hunk 2 above)
Comment (ank27) [reply LICENSE.md]: these are now changed after downgrade the plugin version to `0.8.80`

File: .gitignore (hunk 13 above)
Comment (ank27) [reply .gitignore:31]: why do we need to mention `sdk_versions` are ignored, these are used internally by plugin and not for public-facing changes?

File: android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/testing/CarJavaInterfaceChecker.java (hunk 14 above)
Comment (ank27) [reply android-auto-app/src/main/java/com/mapbox/maps/testapp/auto/testing/CarJavaInterfaceChecker.java:18]: commented same here : https://github.com/mapbox/mapbox-maps-android/pull/1118#discussion_r807803076

File: gradle/lint.gradle (hunk 5 above)
Comment (ank27) [reply gradle/lint.gradle:6]: correct. there are some unused resources. i will try to remove those.

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/animator/CameraAnimator.kt (hunk 10 above)
Comment (ank27) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/animator/CameraAnimator.kt:14]: Since lint error is related to `ValueAnimator` and `CameraAnimator` class extends that.`supressLint` needs to be at the class level.

File: gradle/lint.gradle (hunk 5 above)
Comment (ank27) [reply gradle/lint.gradle:6]: We have quite a few unused resources and some resources that are defined in one module and used in another module. since we run lint on every module, it couldn't figure out if these resources are used in other modules as well. I have created an internal ticket to resolve this, we can do that after this pr is merged. wdyt cc @kiryldz @pengdev ?

File: extension-style/api/extension-style.api (hunk 8 above)
Comment (pengdev) [reply extension-style/api/extension-style.api:3414]: The TileSet extends a `HashMap<String, Value>`.

The difference seems to be that the previous API, the type is erased, while the new API keeps the type.
This might be a compiler behaviour change. And I think the `com/mapbox/bindgen/Value` should be the correct type that java users could safely reference. I think we could consider this as a bug fix and not an API breaking change, wdyt?

File: .gitignore (hunk 13 above)
Comment (kiryldz) [reply .gitignore:31]: yeah, agree about `sdk_versions`

File: extension-style/api/extension-style.api (hunk 8 above)
Comment (ank27) [reply extension-style/api/extension-style.api:3414]: sounds good to me 👍

File: gradle/lint.gradle (hunk 5 above)
Comment (pengdev) [reply gradle/lint.gradle:6]: Yes, since there's a lot of warnings across the codebase, let's disable `'UnusedResources'` and ticket out the tailwork to optimise separately.

File: scripts/java-api-check.json [hunk 15]
```diff
@@ -10,7 +10,56 @@
       "filter": {
         "classes": {
           "regex": true,
-          "include": ["com\\.mapbox\\.maps\\..*"]
+          "include": ["com\\.mapbox\\.maps\\..*"],
```
Comment (kiryldz) [inline scripts/java-api-check.json:13]: @ank27 is it left here intentionally? I guess better to deliver it in a separate PR

**Review by kiryldz — APPROVED** [review]
LGTM, let's merge and iterate

File: sdk/api/sdk.api (hunk 11 above)
Comment (ank27) [reply sdk/api/sdk.api:495]: since its internal class changes, we can skip these as discussed offline.

File: scripts/java-api-check.json (hunk 15 above)
Comment (ank27) [reply scripts/java-api-check.json:13]: yeah, removing this commit.

File: gradle/sdk-registry.gradle [hunk 16]
```diff
@@ -58,13 +58,13 @@ task androidJavadocs(type: Javadoc) {
...
-    classifier = 'javadoc'
+    classifier "javadoc"
     from dokkaJavadoc.outputDirectory
 }
 
 task androidSourcesJar(type: Jar) {
-    classifier = 'sources'
-    from android.sourceSets.main.kotlin
+    classifier "sources"
+    from android.sourceSets.main.java.srcDirs
```
Comment (pengdev) [inline gradle/sdk-registry.gradle:67]: could you verify the android sources jar is generated correctly with this change?

Comment (pengdev) [reply gradle/sdk-registry.gradle:67]: Verified the kotlin dir is picked up correctly for viewport plugin, this should be ok.

**Review by pengdev — APPROVED** [review]
Let's merge and address the remaining issues/tickets as follow up.

File: DEVELOPING.md [hunk 17]
```diff
@@ -30,10 +30,14 @@ The only requirement is to install the latest version of
 [`sdkmanager`](https://developer.android.com/studio/command-line/sdkmanager) command line or
 through the GUI interface to make sure the extra requirements to build Android are met.
 
+Note: We use Android gradle plugin v7.0+ to build the project, which requires JDK 11.
```
Comment (kiryldz) [inline DEVELOPING.md:33]:
```suggestion
Note: Android Gradle Plugin v7.0+ is used to build the project. That requires to have JDK 11 or higher.
```

File: DEVELOPING.md [hunk 18]
```diff
@@ -30,10 +30,14 @@ The only requirement is to install the latest version of
...
+when building project, make sure you have JDK 11 installed and set it as JAVA_HOME variable.
```
Comment (kiryldz) [inline DEVELOPING.md:34]: I'd drop this line at all and use my suggestion above.

**Comment by yunikkk** [comment]
Noting that with AGP7+ compose app could now be a module instead of separate project inside https://github.com/mapbox/mapbox-maps-android/pull/695
